### PR TITLE
Improved UNet Image Segmentation- Tremain Wales 44780164

### PR DESCRIPTION
**Pull request by SirTremain**
This pull request will add 2 files, improvedUNet.py and UNetDriver.py to the folder recognition.

It is an implementation of an improved UNet model used for Image Segmentation. It only depends on tensorflow 2.3 and python 3.7.

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
- Work submitted before deadline 
- Commits not very spaced out. Initial commit is too large showing limited sign of incremental code development.
- Code appear valid for Easy Difficulty, though no mention/demo of results

**Comment by rexsw** [comment]
code appears invalid as outlined above.

For next time;

Code in files like driver scripts should be in a main function

mildly incorrect use of the tensorflow dataset class, you want to use a dataset as a generator not just a way to index the data

File: recognition/44780164_improvedUNet/UNetDriver.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import tensorflow as tf
```
Comment (shakes76) [inline recognition/44780164_improvedUNet/UNetDriver.py:1]: Code header block should be added for all scripts/source files

File: recognition/44780164_improvedUNet/improvedUNetModel.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    segment3 = tf.keras.layers.Conv2D(2, (1, 1), activation=leakyReLU, **CONV_ARGUMENTS)(context7)
+
+    segment3_connector = tf.keras.layers.UpSampling2D(size=(2, 2))(segment3)
+    segment2 = tf.keras.layers.Add()([segment3_connector, segment2])
+    segment2 = tf.keras.layers.UpSampling2D(size=(2, 2))(segment2)
+
+    segment_out = tf.keras.layers.Add()([segment1, segment2])
+
+    # Output
+    final = tf.keras.layers.Conv2D(1, (1, 1), activation='softmax')(segment_out)
```
Comment (shakes76) [inline recognition/44780164_improvedUNet/improvedUNetModel.py:134]: Incorrect use of softmax

File: recognition/44780164_improvedUNet/UNetDriver.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+image_data = data_combination(images_list, masks_list)
+test_dataset = image_data.take(1000)
+remaining = image_data.skip(1000)
+val_dataset = remaining.take(500)
+train_dataset = remaining.skip(500)
+
+model = improvedUNetModel.improvedUNet()
+model.fit(train_dataset.batch(16), epochs=8, validation_data=val_dataset.batch(16))
+
+model.evaluate(test_dataset.bacth(16))
```
Comment (shakes76) [inline recognition/44780164_improvedUNet/UNetDriver.py:62]: batch misspelt, can't be working.

**Comment by shakes76** [comment]
## TF Usage
Not good
## Good Practice (Design/Commenting)
OK design and comments
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage, Readme sparse
Module present
## Commit Log
Meaningful commit messages 
Progressive commits used 
## Documentation
Description and Comments OK
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
**Feedback required. Fix the spelling errors and other issues in the main scripts.**
Fixing the above will only result in obtaining additional feedback marks according to the assessment sheet.
No Request Description. In the future, a more detailed pull request could've been done. When developer receiving the request is not known, the request could be rejected. Remember you could use/reuse the README info.
